examples/scalajs-react/src/main/scala/todomvc/CTodoList.scala

```diff
@@ -0,0 +1,148 @@
+package todomvc
+
+import japgolly.scalajs.react.ScalazReact._
+import japgolly.scalajs.react._
+import japgolly.scalajs.react.extra._
+import japgolly.scalajs.react.extra.router2.RouterCtl
+import japgolly.scalajs.react.vdom.prefix_<^._
+import org.scalajs.dom.ext.KeyCode
+
+import scala.scalajs.js
+import scalaz.effect.IO
+import scalaz.std.anyVal.unitInstance
+import scalaz.syntax.semigroup._
+
+object CTodoList {
+
+  case class Props private[CTodoList] (ctl: RouterCtl[TodoFilter], model: TodoModel, currentFilter: TodoFilter)
+
+  case class State(todos: Seq[Todo], editing: Option[TodoId])
+
+  /**
+   * These specify when it makes sense to skip updating this component (see comment on `Listenable` below)
+   */
+  implicit val r1 = Reusability.fn[Props]((p1, p2) ⇒ p1.currentFilter == p2.currentFilter)
+  implicit val r2 = Reusability.fn[State]((s1, s2) ⇒ s1.editing == s2.editing && (s1.todos eq s2.todos))
+
+  /**
+   * One difference between normal react and scalajs-react is the use of backends.
+   * Since components are not inheritance-based, we often use a backend class
+   * where we put most of the functionality: rendering, state handling, etc.
+   *
+   * It extends OnUnmount so unsubscription of events can be made automatically.
+   */
+  case class Backend($: BackendScope[Props, State]) extends OnUnmount {
+
+    def handleNewTodoKeyDown(event: ReactKeyboardEventI): Option[IO[Unit]] =
+      Some((event.nativeEvent.keyCode, UnfinishedTitle(event.target.value).validated)) collect {
+        case (KeyCode.Enter, Some(title)) =>
+          IO(event.target.value = "") |+| $.props.model.addTodo(title)
+      }
+    
+    def updateTitle(id: TodoId)(title: Title): IO[Unit] =
+      editingDone(cb = $.props.model.update(id, title))
+
+    def startEditing(id: TodoId): IO[Unit] =
+      $.modStateIO(_.copy(editing = Some(id)))
+
+    /**
+     * @param cb Two changes to the same `State` must be combined using a callback like this.
+     *           If not, rerendering will prohibit the second from having its effect.
+     *           For this example, the current `State` contains both `editing` and the list of todos.
+     */
+    def editingDone(cb: OpCallbackIO = js.undefined): IO[Unit] =
```

Comment:
It's actually perfectly typed - `OpCallbackIO` is a type alias defined like this:
```scala
type OpCallbackIO = js.UndefOr[scalaz.effect.IO[Unit]]
```
In other words, it is an optional callback. `js.undefined` can fill that hole in the same way as does `None` for the `Option` type - it's the instance of `js.UndefOr[Nothing]`.

To a pretty big extent we use `js.UndefOr[A]` instead of `Option[A]` in scala.js. It has `Option`-like semantics, but instead uses `js.undefined` as a `null` value instead of runtime wrapper objects. This is done primarily for integration with javascript components. A nice added bonus in the context of react and scala.js is that it also allows us to compare optionals by reference - which you frequently want to do in `shouldComponentUpdate`-